src/Core/Grid/Definition/Factory/MerchandiseReturnProductsGridDefinitionFactory.php

```diff
@@ -0,0 +1,255 @@
+                ],
+            ])
+            ->setAssociatedColumn('product_reference')
+            )
+            ->add((new Filter('product_name', TextType::class))
+            ->setTypeOptions([
+                'required' => false,
+                'attr' => [
+                    'placeholder' => $this->trans('Search name', [], 'Admin.Actions'),
+                ],
+            ])
+            ->setAssociatedColumn('product_name')
+            )
+            ->add((new Filter('quantity', TextType::class))
+            ->setTypeOptions([
+                'required' => false,
+                'attr' => [
+                    'placeholder' => $this->trans('Search quantity', [], 'Admin.Actions'),
+                ],
+            ])
+            ->setAssociatedColumn('quantity')
+            )
+            ->add((new Filter('actions', SearchAndResetType::class))
+            ->setTypeOptions([
+                'reset_route' => 'admin_common_reset_search_by_filter_id',
+                'reset_route_params' => [
+                    'filterId' => self::GRID_ID,
+                    'merchandiseReturnId' => $this->merchandiseReturnId,
+                ],
+                'redirect_route' => 'admin_merchandise_returns_edit',
+                'redirect_route_params' => [
+                    'merchandiseReturnId' => $this->merchandiseReturnId,
+                ],
+            ])
+            ->setAssociatedColumn('actions')
+            )
+        ;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getBulkActions()
+    {
+        return (new BulkActionCollection())
+            ->add(
+                (new SubmitBulkAction('delete_selection'))
+                    ->setName($this->trans('Delete selection', [], 'Admin.Actions'))
+                    ->setOptions([
+                        'submit_route' => 'admin_merchandise_returns_delete_product_bulk',
+                        'route_params' => [
+                            'merchandiseReturnId' => $this->merchandiseReturnId,
+                        ],
+                        'confirm_message' => $this->trans('Are you sure you want to delete the selected item(s)?', [], 'Admin.Global'),
+                        'modal_options' => new ModalOptions([
+                            'title' => $this->trans('Delete selection', [], 'Admin.Actions'),
+                            'confirm_button_label' => $this->trans('Delete', [], 'Admin.Actions'),
+                            'confirm_button_class' => 'btn-danger',
+                        ]),
+                    ])
+            )
+            ;
+    }
+
+    /**
+     * Sets cms page category parent id directly from request attribute. On not found case, it assigns the default one.
+     *
+     * @param RequestStack $requestStack
+     *
+     * @throws MerchandiseReturnProductException
+     */
+    private function setMerchandiseReturnId(RequestStack $requestStack): void
```

Comment:
Herr, this one using RequestStack is a bit ugly as well 😅  like what we did for the listener But I have no alternative right now We'll need to think about how we could improve this as well 🤔

Reply:
Yeah at the moment I couldn't think of any better ideas, so reused what was already used.